### PR TITLE
Allow markdown to match index file for extensions.  Fix #1418

### LINKED ISSUE #1418: Markdown does not apply to default directory index file
(Are you asking for help with using Caddy? Please use our forum instead: https://forum.caddyserver.com. If you are filing a bug report, please answer the following questions. If your issue is not a bug report, you do not need to use this template. Either way, please consider donating if we've helped you. Thanks!)

### 1. What version of Caddy are you running (`caddy -version`)?
0.9.5

### 2. What are you trying to do?

Use .html as the markdown extension and have file load as index file for directory with markdown

### 3. What is your entire Caddyfile?
aubergine.kinchik.ie:80 {
    root d:\caddy\webroot\mdissue
    markdown /md {
        ext .html
    }

}

### 4. How did you run Caddy (give the full command and describe the execution environment)?

caddy -conf=d:\godev\bin\caddymd.txt -log stdout
### 5. What did you expect to see?

when i visit
> http://aubergine.kinchik.ie/md/

I expect to see http://aubergine.kinchik.ie/md/index.html  rendered in markdown

### 6. What did you see instead (give full error messages and/or log)?

http://aubergine.kinchik.ie/md/index.html rendered as html 

If I visit http://aubergine.kinchik.ie/md/index.html directly then I see the markdown rendered properly, but not when the file is omitted.

Should there be a index directive like there is for the fastcgi and other directives.  This would also remove the necessity to use .html as a md ext

### 7. How can someone who is starting from scratch reproduce this behavior as minimally as possible?

Use the caddyfile provided.

### PR DESCRIPTION
**Pull request by tobya**
This PR ensures that an index file for each markdown extension is check for. Fixes #1418 

Edit: I realised I had explained this wrong.

The only index file that was getting matched for markdown was index.md even if the extension was different.

This PR adds an index file to be matched for each extension provided for markdown.  Index files are now also matched on a block basis.

As ever my newness to GO may betray me so I am open to any comments about what changes I should make to structure my code in a better way.

The default extensions (.md, .markdown and .mdown) are represented as a hash rather than array so  if an index.md, index.markdown and index.mdown exist in a directory it is random as to which gets displayed.  I decide this was unlikely to happen so ok to leave.

### CODE REVIEW

File: caddyhttp/markdown/markdown.go [hunk 1]
```diff
@@ -107,6 +107,7 @@ func (md Markdown) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error
 
 		// Set path to found index file
 		fpath = idx
+
```
Comment (mholt) [inline caddyhttp/markdown/markdown.go]: Not a showstopper, but in the future, I prefer that closing curly braces don't have extra blank lines before them, it makes it easier to see exactly what it's enclosing.

File: caddyhttp/markdown/setup.go [hunk 2]
```diff
@@ -75,6 +75,14 @@ func markdownParse(c *caddy.Controller) ([]*Config, error) {
...
 		}
 
+		// Make a list of index files to match extensions
+		i := 0
+		md.IndexFiles = make([]string, len(md.Extensions))
+		for ext := range md.Extensions {
+			indexName := "index" + ext
+			md.IndexFiles[i] = indexName
+			i++
+		}
```
Comment (mholt) [inline caddyhttp/markdown/setup.go]: I think this whole change can be simplified:

```go
for ext := range md.Extensions {
    md.IndexFiles = append(md.IndexFiles, "index" + ext)
}
```

Pre-allocating is good for a slice that will be very large, but when there's like 5 or 10 things, I prefer shorter, less error-prone code.

Comment (tobya) [inline caddyhttp/markdown/setup.go]: @mholt understood. I've made that change now

**Comment by mholt** [comment]
Fabulous - thanks again Toby!

**Pull request merged**